File: backend/src/main/java/com/tailorstudio/app/service/OtpAuthService.java
package com.tailorstudio.app.service;

import com.tailorstudio.app.domain.OtpChallenge;
import com.tailorstudio.app.domain.OtpPurpose;
import com.tailorstudio.app.domain.PasswordResetToken;
import com.tailorstudio.app.domain.PendingLogin;
import com.tailorstudio.app.domain.User;
import com.tailorstudio.app.mail.OtpMailDispatchEvent;
import com.tailorstudio.app.mail.StudioMailSender;
import com.tailorstudio.app.mail.StudioOtpEmailHtml;
import com.tailorstudio.app.repo.OtpChallengeRepository;
import com.tailorstudio.app.repo.PasswordResetTokenRepository;
import com.tailorstudio.app.repo.PendingLoginRepository;
import com.tailorstudio.app.repo.UserRepository;
import com.tailorstudio.app.security.JwtService;
import com.tailorstudio.app.security.OtpCodeHasher;
import jakarta.annotation.PostConstruct;
import org.bson.types.ObjectId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Locale;

@Service
public class OtpAuthService {

    private static final Logger log = LoggerFactory.getLogger(OtpAuthService.class);

    private static final Duration OTP_TTL = Duration.ofMinutes(2);
    private static final Duration PENDING_LOGIN_TTL = Duration.ofMinutes(10);
    private static final Duration RESET_TOKEN_TTL = Duration.ofMinutes(15);

    public record LoginVerifyOutcome(User user, String accessToken) {}

    private final UserRepository userRepository;
    private final OtpChallengeRepository otpChallengeRepository;
    private final PasswordResetTokenRepository passwordResetTokenRepository;
    private final PendingLoginRepository pendingLoginRepository;
    private final AuthService authService;
    private final PasswordEncoder passwordEncoder;
    private final StudioMailSender studioMailSender;
    private final UserAuthLookup userAuthLookup;
    private final ApplicationEventPublisher eventPublisher;
    private final UserSessionEpochService userSessionEpochService;
    private final JwtService jwtService;
    private final String otpPepper;
    private final String brandPublicUrl;
    private final boolean staticOtpEnabledRaw;
    private final String staticOtpCodeRaw;
    private boolean staticOtpActive;
    /** Exactly six digits from {@link #staticOtpCodeRaw}, used for comparison after {@link #initStaticOtp}. */
    private String staticOtpDigits = "";

    public OtpAuthService(
            UserRepository userRepository,
            OtpChallengeRepository otpChallengeRepository,
            PasswordResetTokenRepository passwordResetTokenRepository,
            PendingLoginRepository pendingLoginRepository,
            AuthService authService,
            PasswordEncoder passwordEncoder,
            StudioMailSender studioMailSender,
            UserAuthLookup userAuthLookup,
            ApplicationEventPublisher eventPublisher,
            UserSessionEpochService userSessionEpochService,
            JwtService jwtService,
            @Value("${app.otp.pepper}") String otpPepper,
            @Value("${app.brand.public-url:}") String brandPublicUrl,
            @Value("${app.otp.static-enabled:false}") boolean staticOtpEnabledRaw,
            @Value("${app.otp.static-code:}") String staticOtpCodeRaw) {
        this.userRepository = userRepository;
        this.otpChallengeRepository = otpChallengeRepository;
        this.passwordResetTokenRepository = passwordResetTokenRepository;
        this.pendingLoginRepository = pendingLoginRepository;
        this.authService = authService;
        this.passwordEncoder = passwordEncoder;
        this.studioMailSender = studioMailSender;
        this.userAuthLookup = userAuthLookup;
        this.eventPublisher = eventPublisher;
        this.userSessionEpochService = userSessionEpochService;
        this.jwtService = jwtService;
        this.otpPepper = otpPepper;
        this.brandPublicUrl = brandPublicUrl;
        this.staticOtpEnabledRaw = staticOtpEnabledRaw;
        this.staticOtpCodeRaw = staticOtpCodeRaw;
    }

    @PostConstruct
    void initStaticOtp() {
        if (!staticOtpEnabledRaw) {
            staticOtpActive = false;
            staticOtpDigits = "";
            return;
        }
        String digits = staticOtpCodeRaw == null ? "" : staticOtpCodeRaw.replaceAll("\\D", "");
        if (!digits.matches("\\d{6}")) {
            log.error(
                    "STATIC_OTP_ENABLED is true but STATIC_OTP / app.otp.static-code is not exactly 6 digits — static OTP bypass disabled.");
            staticOtpActive = false;
            staticOtpDigits = "";
            return;
        }
        staticOtpDigits = digits;
        staticOtpActive = true;
        log.warn("STATIC OTP bypass is ON (code length 6). Do not use in production with real users.");
    }

    private boolean staticOtpBypass() {
        return staticOtpActive;
    }

    /** Exposed for login challenge JSON so the client can skip time-based OTP expiry UX when mail OTP is bypassed. */
    public boolean isStaticOtpBypassEnabled() {
        return staticOtpBypass();
    }

    private boolean isSubmittedStaticOtp(String codeRaw) {
        if (!staticOtpBypass() || codeRaw == null) {
            return false;
        }
        String submitted = codeRaw.replaceAll("\\D", "");
        return staticOtpDigits.equals(submitted);
    }

    public boolean isMailConfigured() {
        return studioMailSender.isConfigured();
    }

    private void requireMailUnlessStaticBypass() {
        if (staticOtpBypass()) {
            return;
        }
        if (!studioMailSender.isConfigured()) {
            throw new IllegalStateException("Email is not configured (set EMAIL and EMAIL_PASSWORD).");
        }
    }

    private void enqueueOtpEmail(String to, String subject, String plainText, String htmlBody) {
        eventPublisher.publishEvent(new OtpMailDispatchEvent(to, subject, plainText, htmlBody));
    }

    private String normalizeEmail(String email) {
        return email == null ? "" : email.trim().toLowerCase(Locale.ROOT);
    }

    public record LoginChallengeResult(Instant expiresAt, String pendingTokenPlain) {}

    @Transactional
    public LoginChallengeResult startLoginWithPassword(String emailRaw, String passwordRaw) {
        requireMailUnlessStaticBypass();
        User user = userAuthLookup.findByEmailFlexible(emailRaw).orElse(null);
        if (user == null) {
            throw new NoSuchUserException();
        }
        if (!passwordEncoder.matches(passwordRaw, user.getPasswordHash())) {
            throw new BadPasswordException();
        }
        String email = normalizeEmail(user.getEmail());
        pendingLoginRepository.deleteByEmailIgnoreCase(email);
        String pendingTokenPlain = OtpCodeHasher.randomTokenHex(24);
        PendingLogin pl = new PendingLogin();
        pl.setId(new ObjectId().toHexString());
        pl.setEmail(email);
        pl.setUserId(user.getId());
        pl.setTokenHash(OtpCodeHasher.sha256Hex(pendingTokenPlain));
        pl.setExpiresAt(Instant.now().plus(PENDING_LOGIN_TTL));
        pendingLoginRepository.save(pl);
        Instant expiresAt = Instant.now().plus(OTP_TTL);
        if (!staticOtpBypass()) {
            String code = OtpCodeHasher.newSixDigitOtp();
            OtpChallenge c = persistChallenge(email, OtpPurpose.LOGIN, code);
            expiresAt = c.getExpiresAt();
            enqueueOtpEmail(
                    user.getEmail(),
                    "Tailor Studio — sign-in code",
                    StudioOtpEmailHtml.loginPlain(code),
                    StudioOtpEmailHtml.loginEmail(code, brandPublicUrl));
        }
        return new LoginChallengeResult(expiresAt.truncatedTo(ChronoUnit.MILLIS), pendingTokenPlain);
    }

    @Transactional
    public Instant resendLoginOtp(String pendingTokenPlain) {
        requireMailUnlessStaticBypass();
        String th = OtpCodeHasher.sha256Hex(pendingTokenPlain.trim());
        PendingLogin pl = pendingLoginRepository.findByTokenHashAndExpiresAtAfter(th, Instant.now()).orElseThrow(OtpInvalidException::new);
        User user = userRepository.findById(pl.getUserId()).orElseThrow(OtpInvalidException::new);
        if (!normalizeEmail(user.getEmail()).equals(normalizeEmail(pl.getEmail()))) {
            throw new OtpInvalidException();
        }
        String email = normalizeEmail(pl.getEmail());
        if (staticOtpBypass()) {
            return Instant.now().plus(OTP_TTL);
        }
        String code = OtpCodeHasher.newSixDigitOtp();
        OtpChallenge c = persistChallenge(email, OtpPurpose.LOGIN, code);
        enqueueOtpEmail(
                user.getEmail(),
                "Tailor Studio — sign-in code",
                StudioOtpEmailHtml.loginPlain(code),
                StudioOtpEmailHtml.loginEmail(code, brandPublicUrl));
        return c.getExpiresAt();
    }

    @Transactional
    public Instant sendForgotPasswordOtp(String emailRaw) {
        requireMailUnlessStaticBypass();
        User user = userAuthLookup.findByEmailFlexible(emailRaw).orElse(null);
        if (user == null) {
            throw new NoSuchUserException();
        }
        String email = normalizeEmail(user.getEmail());
        if (staticOtpBypass()) {
            return Instant.now().plus(OTP_TTL);
        }
        String code = OtpCodeHasher.newSixDigitOtp();
        OtpChallenge c = persistChallenge(email, OtpPurpose.PASSWORD_RESET, code);
        enqueueOtpEmail(
                user.getEmail(),
                "Tailor Studio — reset password",
                StudioOtpEmailHtml.passwordResetPlain(code),
                StudioOtpEmailHtml.passwordResetEmail(code, brandPublicUrl));
        return c.getExpiresAt();
    }

    private OtpChallenge persistChallenge(String email, OtpPurpose purpose, String code) {
        otpChallengeRepository.deleteByEmailIgnoreCaseAndPurpose(email, purpose);
        OtpChallenge c = new OtpChallenge();
        c.setId(new ObjectId().toHexString());
        c.setEmail(email);
        c.setPurpose(purpose);
        c.setCodeHash(OtpCodeHasher.hashOtp(otpPepper, email, code));
        c.setExpiresAt(Instant.now().plus(OTP_TTL));
        return otpChallengeRepository.save(c);
    }

    @Transactional
    public LoginVerifyOutcome verifyLoginOtp(String emailRaw, String code, String pendingTokenPlain) {
        String email = normalizeEmail(emailRaw);
        String th = OtpCodeHasher.sha256Hex(pendingTokenPlain.trim());
        PendingLogin pl = pendingLoginRepository.findByTokenHashAndExpiresAtAfter(th, Instant.now()).orElseThrow(OtpInvalidException::new);
        if (!normalizeEmail(pl.getEmail()).equals(email)) {
            throw new OtpInvalidException();
        }
        Instant now = Instant.now();
        if (isSubmittedStaticOtp(code)) {
            otpChallengeRepository.deleteByEmailIgnoreCaseAndPurpose(email, OtpPurpose.LOGIN);
            pendingLoginRepository.deleteById(pl.getId());
            User user = userRepository.findById(pl.getUserId()).orElseThrow();
            return issueLoginTokens(user);
        }
        OtpChallenge c = otpChallengeRepository
                .findTopByEmailIgnoreCaseAndPurposeAndExpiresAtAfterOrderByCreatedAtDesc(email, OtpPurpose.LOGIN, now)
                .orElse(null);
        if (c == null) {
            throw new OtpInvalidException();
        }
        String want = c.getCodeHash();
        String got = OtpCodeHasher.hashOtp(otpPepper, email, code.trim());
        if (!OtpCodeHasher.constantTimeEquals(want, got)) {
            throw new OtpInvalidException();
        }
        otpChallengeRepository.deleteById(c.getId());
        pendingLoginRepository.deleteById(pl.getId());
        User user = userRepository.findById(pl.getUserId()).orElseThrow();
        return issueLoginTokens(user);
    }

    private LoginVerifyOutcome issueLoginTokens(User user) {
        long epoch = userSessionEpochService.bumpEpoch(user.getId());
        User refreshed = userRepository.findById(user.getId()).orElse(user);
        String accessToken = jwtService.createAccessToken(refreshed.getId(), epoch);
        return new LoginVerifyOutcome(refreshed, accessToken);
    }

    @Transactional
    public String verifyForgotOtpAndIssueResetToken(String emailRaw, String code) {
        String email = normalizeEmail(emailRaw);
        if (isSubmittedStaticOtp(code)) {
            User user = userAuthLookup.findByEmailFlexible(emailRaw).orElseThrow(OtpInvalidException::new);
            if (!normalizeEmail(user.getEmail()).equals(normalizeEmail(emailRaw))) {
                throw new OtpInvalidException();
            }
            otpChallengeRepository.deleteByEmailIgnoreCaseAndPurpose(email, OtpPurpose.PASSWORD_RESET);
            passwordResetTokenRepository.deleteByUserId(user.getId());
            String token = OtpCodeHasher.randomTokenHex(24);
            PasswordResetToken pr = new PasswordResetToken();
            pr.setId(new ObjectId().toHexString());
            pr.setUserId(user.getId());
            pr.setTokenHash(OtpCodeHasher.sha256Hex(token));
            pr.setExpiresAt(Instant.now().plus(RESET_TOKEN_TTL));
            passwordResetTokenRepository.save(pr);
            return token;
        }
        Instant now = Instant.now();
        OtpChallenge c = otpChallengeRepository
                .findTopByEmailIgnoreCaseAndPurposeAndExpiresAtAfterOrderByCreatedAtDesc(email, OtpPurpose.PASSWORD_RESET, now)
                .orElse(null);
        if (c == null) {
            throw new OtpInvalidException();
        }
        String want = c.getCodeHash();
        String got = OtpCodeHasher.hashOtp(otpPepper, email, code.trim());
        if (!OtpCodeHasher.constantTimeEquals(want, got)) {
            throw new OtpInvalidException();
        }
        otpChallengeRepository.deleteById(c.getId());
        User user = userAuthLookup.findByEmailFlexible(emailRaw).orElseThrow();
        passwordResetTokenRepository.deleteByUserId(user.getId());
        String token = OtpCodeHasher.randomTokenHex(24);
        PasswordResetToken pr = new PasswordResetToken();
        pr.setId(new ObjectId().toHexString());
        pr.setUserId(user.getId());
        pr.setTokenHash(OtpCodeHasher.sha256Hex(token));
        pr.setExpiresAt(Instant.now().plus(RESET_TOKEN_TTL));
        passwordResetTokenRepository.save(pr);
        return token;
    }

    @Transactional
    public void resetPasswordWithToken(String tokenRaw, String newPassword) {
        if (newPassword == null || newPassword.length() < 8) {
            throw new IllegalArgumentException("Password must be at least 8 characters");
        }
        String th = OtpCodeHasher.sha256Hex(tokenRaw.trim());
        PasswordResetToken pr = passwordResetTokenRepository.findByTokenHashAndExpiresAtAfter(th, Instant.now()).orElseThrow(OtpInvalidException::new);
        authService.updatePasswordForUser(pr.getUserId(), newPassword);
        passwordResetTokenRepository.deleteById(pr.getId());
    }

    public static final class NoSuchUserException extends RuntimeException {}

    public static final class BadPasswordException extends RuntimeException {}

    public static final class OtpInvalidException extends RuntimeException {}
}

File: server/src/services/otpAuthService.ts
import bcrypt from 'bcryptjs'
import { OtpChallenge, PendingLogin, PasswordResetToken, User } from '../models/index.js'
import { env } from '../config/env.js'
import { isMailConfigured, sendMailAsync } from '../mail/mailer.js'
import * as mail from '../mail/otpEmail.js'
import {
  constantTimeEquals,
  hashOtp,
  isStaticOtp,
  newSixDigitOtp,
  randomTokenHex,
  sha256Hex,
} from '../utils/otp.js'
import { createAccessToken } from '../utils/jwt.js'
import { findByEmailFlexible } from './userAuthLookup.js'
import { updatePasswordForUser } from './authService.js'
import { parseObjectId } from '../utils/objectId.js'
import type { UserDoc } from '../models/User.js'
import { userFromLean } from '../models/User.js'
import { leanOne } from '../utils/lean.js'

const OTP_MS = 120_000
const RESET_MS = 900_000
const norm = (e: string) => e.trim().toLowerCase()

function escapeRegex(s: string) {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

function requireMail() {
  if (!env.staticOtpEnabled && !isMailConfigured()) {
    throw new Error('Email is not configured (set EMAIL and EMAIL_PASSWORD).')
  }
}

async function saveChallenge(email: string, purpose: string, code: string) {
  const emailKey = norm(email)
  await OtpChallenge.deleteMany({ email: emailKey, purpose })
  const expiresAt = new Date(Date.now() + OTP_MS)
  await OtpChallenge.create({ email: emailKey, purpose, codeHash: hashOtp(emailKey, code), expiresAt })
  return expiresAt
}

async function findLoginChallenge(email: string) {
  const emailKey = norm(email)
  return OtpChallenge.findOne({
    email: { $regex: new RegExp(`^${escapeRegex(emailKey)}$`, 'i') },
    purpose: 'LOGIN',
    expiresAt: { $gt: new Date() },
  })
    .sort({ createdAt: -1 })
    .lean()
}

async function assertCredentials(emailRaw: string, password: string): Promise<UserDoc> {
  const user = await findByEmailFlexible(emailRaw)
  if (!user) throw new Error('no_account')
  if (!(await bcrypt.compare(password, user.passwordHash))) throw new Error('invalid_credentials')
  return user
}

function issueLoginSession(user: UserDoc) {
  if (!user.id) throw new Error('user_not_found')
  return { user, accessToken: createAccessToken(user.id) }
}

async function checkLoginOtpForUser(user: UserDoc, codeRaw: unknown) {
  const email = norm(user.email)
  const digits = String(codeRaw ?? '').replace(/\D/g, '')
  if (digits.length !== 6) throw new Error('invalid_otp')

  if (isStaticOtp(digits)) {
    await OtpChallenge.deleteMany({ email, purpose: 'LOGIN' })
    return issueLoginSession(user)
  }

  const c = await findLoginChallenge(email)
  if (!c || !constantTimeEquals(leanOne<{ codeHash: string }>(c).codeHash, hashOtp(email, digits))) {
    throw new Error('invalid_otp')
  }
  await OtpChallenge.deleteOne({ _id: leanOne<{ _id: unknown }>(c)._id })
  return issueLoginSession(user)
}

export { isMailConfigured }
export const isStaticOtpBypassEnabled = () => env.staticOtpEnabled

export async function startLoginWithPassword(emailRaw: string, password: string) {
  requireMail()
  const user = await assertCredentials(emailRaw, password)
  const email = norm(user.email)
  let expiresAt = new Date(Date.now() + OTP_MS)
  if (env.staticOtpEnabled) {
    expiresAt = await saveChallenge(email, 'LOGIN', env.staticOtp)
  } else {
    const code = newSixDigitOtp()
    expiresAt = await saveChallenge(email, 'LOGIN', code)
    sendMailAsync({
      to: user.email,
      subject: 'Tailor Studio — sign-in code',
      text: mail.loginPlain(code),
      html: mail.loginEmail(code),
    })
  }
  return { expiresAt, email: user.email }
}

export async function completeLoginWithPasswordAndOtp(
  emailRaw: string,
  password: string,
  codeRaw: unknown,
) {
  const user = await assertCredentials(emailRaw, password)
  return checkLoginOtpForUser(user, codeRaw)
}

export async function resendLoginOtpWithPassword(emailRaw: string, password: string) {
  requireMail()
  const user = await assertCredentials(emailRaw, password)
  const email = norm(user.email)
  if (env.staticOtpEnabled) {
    return saveChallenge(email, 'LOGIN', env.staticOtp)
  }
  const code = newSixDigitOtp()
  const expiresAt = await saveChallenge(email, 'LOGIN', code)
  sendMailAsync({
    to: user.email,
    subject: 'Tailor Studio — sign-in code',
    text: mail.loginPlain(code),
    html: mail.loginEmail(code),
  })
  return expiresAt
}

export async function resendLoginOtp(pendingTokenPlain: string) {
  requireMail()
  const pl = await PendingLogin.findOne({
    tokenHash: sha256Hex(pendingTokenPlain.trim()),
    expiresAt: { $gt: new Date() },
  }).lean()
  if (!pl) throw new Error('invalid_pending')
  const plRow = leanOne<{ userId: unknown; email: string }>(pl)
  const raw = await User.findById(parseObjectId(String(plRow.userId))).lean()
  if (!raw || norm(leanOne<{ email: string }>(raw).email) !== norm(plRow.email)) {
    throw new Error('invalid_pending')
  }
  const user = userFromLean(leanOne<Record<string, unknown>>(raw))
  const email = norm(plRow.email)
  if (env.staticOtpEnabled) {
    return saveChallenge(email, 'LOGIN', env.staticOtp)
  }
  const code = newSixDigitOtp()
  const expiresAt = await saveChallenge(email, 'LOGIN', code)
  sendMailAsync({
    to: user.email,
    subject: 'Tailor Studio — sign-in code',
    text: mail.loginPlain(code),
    html: mail.loginEmail(code),
  })
  return expiresAt
}

export async function verifyLoginOtp(_emailRaw: string, codeRaw: unknown, pendingTokenPlain: unknown) {
  const token = String(pendingTokenPlain ?? '').trim()
  if (!token) throw new Error('invalid_pending')

  const pl = await PendingLogin.findOne({
    tokenHash: sha256Hex(token),
    expiresAt: { $gt: new Date() },
  }).lean()
  if (!pl) throw new Error('invalid_pending')

  const plRow = leanOne<{ email: string; userId: unknown; _id: unknown }>(pl)
  const raw = await User.findById(parseObjectId(String(plRow.userId))).lean()
  if (!raw) throw new Error('invalid_pending')
  const user = userFromLean(leanOne<Record<string, unknown>>(raw))
  await PendingLogin.deleteOne({ _id: plRow._id })
  return checkLoginOtpForUser(user, codeRaw)
}

export async function sendForgotPasswordOtp(emailRaw: string) {
  requireMail()
  const user = await findByEmailFlexible(emailRaw)
  if (!user) throw new Error('no_account')
  const email = norm(user.email)
  if (env.staticOtpEnabled) {
    return saveChallenge(email, 'PASSWORD_RESET', env.staticOtp)
  }
  const code = newSixDigitOtp()
  const expiresAt = await saveChallenge(email, 'PASSWORD_RESET', code)
  sendMailAsync({
    to: user.email,
    subject: 'Tailor Studio — reset password',
    text: mail.passwordResetPlain(code),
    html: mail.passwordResetEmail(code),
  })
  return expiresAt
}

export async function verifyForgotOtp(emailRaw: string, code: string) {
  const email = norm(emailRaw)
  const user = await findByEmailFlexible(emailRaw)
  if (!user) throw new Error('invalid_otp')

  const issue = async () => {
    await PasswordResetToken.deleteMany({ userId: parseObjectId(user.id) })
    const token = randomTokenHex(24)
    await PasswordResetToken.create({
      userId: parseObjectId(user.id),
      tokenHash: sha256Hex(token),
      expiresAt: new Date(Date.now() + RESET_MS),
    })
    return token
  }

  if (isStaticOtp(code)) {
    await OtpChallenge.deleteMany({ email, purpose: 'PASSWORD_RESET' })
    return issue()
  }

  const c = await OtpChallenge.findOne({
    email,
    purpose: 'PASSWORD_RESET',
    expiresAt: { $gt: new Date() },
  })
    .sort({ createdAt: -1 })
    .lean()
  if (!c || !constantTimeEquals(leanOne<{ codeHash: string }>(c).codeHash, hashOtp(email, code.trim()))) {
    throw new Error('invalid_otp')
  }
  await OtpChallenge.deleteOne({ _id: leanOne<{ _id: unknown }>(c)._id })
  return issue()
}

export async function resetPasswordWithToken(tokenRaw: string, newPassword: string) {
  if (!newPassword || newPassword.length < 8) throw new Error('Password must be at least 8 characters')
  const pr = await PasswordResetToken.findOne({
    tokenHash: sha256Hex(tokenRaw.trim()),
    expiresAt: { $gt: new Date() },
  }).lean()
  if (!pr) throw new Error('invalid_reset_token')
  await updatePasswordForUser(String(leanOne<{ userId: unknown }>(pr).userId), newPassword)
  await PasswordResetToken.deleteOne({ _id: leanOne<{ _id: unknown }>(pr)._id })
}
